### PR TITLE
Feature/ia ddd value objects

### PR DESCRIPTION
**Pull request by tamaragovorcin**

### CODE REVIEW

File: IntegrationWithPharmacies/EPrescriptionApi/AbstractFactory/PharmacyFactoryHttp.cs [hunk 1]
```diff
@@ -0,0 +1,15 @@
+﻿using EPrescriptionApi.DbContextModel;
+using System;
+
+namespace EPrescriptionApi.AbstractFactory
+{
+    public class PharmacyFactoryHttp : AbstractFactory
+    {
+        public override IPharmacy GetIPharmacy(MyDbContext context)
+        {
+            Console.WriteLine("OVDE SAM GAAAAAAAAAAAAAAAAAAAAAA");
```
Comment (trki293) [inline IntegrationWithPharmacies/EPrescriptionApi/AbstractFactory/PharmacyFactoryHttp.cs:10]: Eliminate unnecessary console printing.

File: IntegrationWithPharmacies/EPrescriptionApi/AbstractFactory/PharmacyHttp.cs [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+    {
+        private SftpService SftpService { get; }
+        private HttpRequests HttpRequests { get; }
+        private SmptServerService SmptServerService { get; }
+
+        private PrescriptionFileService PrescriptionFileService { get; }
+        public PharmacyHttp() { }
+        public PharmacyHttp(MyDbContext context)
+        {
+            Console.WriteLine("***************************************          TUUUUUUUUUU");
```
Comment (trki293) [inline IntegrationWithPharmacies/EPrescriptionApi/AbstractFactory/PharmacyHttp.cs:21]: Eliminate unnecessary console printing.

File: IntegrationWithPharmacies/MedicineInformationApi/Service/MedicineInformationService.cs [hunk 3]
```diff
@@ -0,0 +1,109 @@
...
+using System;
+using System.Collections.Generic;
+using System.Linq;
+
+namespace MedicineInformationApi.Service
+{
+    public class MedicineInformationService : IMedicineInformationService
+    {
+        public MedicineInformationRepository MedicineWithQuantityRepository { get; }
+        public IMedicineInformationRepository IMedicineWithQuantityRepository { get; }
```
Comment (trki293) [inline IntegrationWithPharmacies/MedicineInformationApi/Service/MedicineInformationService.cs:15]: Change name of interface IMedicineWithQuantityRepository to _medicineWithQuantityRepository.

File: IntegrationWithPharmacies/MedicineInformationApi/Service/MedicineInformationService.cs [hunk 4]
```diff
@@ -0,0 +1,109 @@
...
+        public MedicineInformationService() { }
+
+        public MedicineInformationService(MyDbContext context)
+        {
+            MedicineWithQuantityRepository = new MedicineInformationRepository(context);
+        }
+        public MedicineInformationService(IMedicineInformationRepository mdicineWithQuantityRepository)
```
Comment (trki293) [inline IntegrationWithPharmacies/MedicineInformationApi/Service/MedicineInformationService.cs:22]: Change name of parameter to medicineWithQuantityRepository

File: IntegrationWithPharmacies/MedicineInformationApi/Service/MedicineInformationService.cs [hunk 5]
```diff
@@ -0,0 +1,109 @@
...
+
+        public List<MedicineInformation> GetAllForStub()
+        {
+            return IMedicineWithQuantityRepository.GetAll();
+        }
+
+        public List<MedicineName>  GetAllMedicinesFromDatabase()
+        {
+            List<MedicineInformation> medicines = MedicineWithQuantityRepository.GetAll();
+            return getMedicineNames(medicines);
```
Comment (trki293) [inline IntegrationWithPharmacies/MedicineInformationApi/Service/MedicineInformationService.cs:52]: Write this in form:
"return getMedicineNames(MedicineWithQuantityRepository.GetAll());"

File: IntegrationWithPharmacies/PharmacyRegistrationApi/Service/RegistrationInPharmacyService.cs [hunk 6]
```diff
@@ -30,14 +30,14 @@ public List<RegistrationInPharmacy> GetAll()
         public RegistrationInPharmacy Create(RegistrationInPharmacyDto dto)
         {
             RegistrationInPharmacy registration = RegistrationInPharmacyAdapter.RegistrationDtoToRegistration(dto);
-            if (isApiKeyUnique(registration.ApiKey))  return RegistrationInPharmacyRepository.Create(registration); 
-             return null;
+            if (isApiKeyUnique(registration.PharmacyConnectionInfo.ApiKey))  return RegistrationInPharmacyRepository.Create(registration); 
+            return null;
```
Comment (trki293) [inline IntegrationWithPharmacies/PharmacyRegistrationApi/Service/RegistrationInPharmacyService.cs:34]: Write this as a ternary operator.

File: IntegrationWithPharmacies/PharmacyRegistrationApi/Service/RegistrationInPharmacyService.cs [hunk 7]
```diff
@@ -30,14 +30,14 @@ public List<RegistrationInPharmacy> GetAll()
...
         }
         public bool isApiKeyUnique(String apiKey)
```
Comment (trki293) [inline IntegrationWithPharmacies/PharmacyRegistrationApi/Service/RegistrationInPharmacyService.cs:36]: Change the method name to start with an initial capital letter.

File: IntegrationWithPharmacies/PharmacyRegistrationApi/Service/RegistrationInPharmacyService.cs [hunk 8]
```diff
@@ -50,23 +50,23 @@ public RegistrationInPharmacy getPharmacyApiKey(String apiKey)
...
         public RegistrationInPharmacy GetRegistrationByPharmacyName(String name)
         {
             foreach (RegistrationInPharmacy registration in RegistrationInPharmacyRepository.GetAll())
             {
-                if (registration.Name.Equals(name)) return registration;
+                if (registration.PharmacyNameInfo.Name.Equals(name)) return registration;
             }
             return null;
         }
         public RegistrationInPharmacy createIRegistration(RegistrationInPharmacyDto dto)
```
Comment (trki293) [inline IntegrationWithPharmacies/PharmacyRegistrationApi/Service/RegistrationInPharmacyService.cs:65]: Change the method name to start with an initial capital letter.

File: IntegrationWithPharmacies/PharmacyRegistrationApi/Service/RegistrationInPharmacyService.cs [hunk 9]
```diff
@@ -50,23 +50,23 @@ public RegistrationInPharmacy getPharmacyApiKey(String apiKey)
         {
```
Comment (trki293) [inline IntegrationWithPharmacies/PharmacyRegistrationApi/Service/RegistrationInPharmacyService.cs:50]: Change the method name 'getPharmacyApiKey' to start with an initial capital letter.

File: IntegrationWithPharmacies/TenderApi/Service/MedicineForTenderingService.cs [hunk 10]
```diff
@@ -47,7 +48,8 @@ public void CreateAllMedicineForTendering(TenderOrder tender)
         {
             foreach (MedicineTenderOffer medicineQuantity in tender.MedicinesWithQuantity)
             {
-                Create(MedicineForTenderingAdapter.MedicineForebderingToMedicineForTenderingDto(new MedicineForTendering(medicineQuantity.MedicineName, medicineQuantity.Quantity, TenderService.getNextTenderId())));
+                Console.WriteLine(medicineQuantity.RequiredQuantity);
```
Comment (trki293) [inline IntegrationWithPharmacies/TenderApi/Service/MedicineForTenderingService.cs:51]: Delete the line if console printing is unnecessary.

File: IntegrationWithPharmacies/TenderApi/Service/MedicineTenderOfferService.cs [hunk 11]
```diff
@@ -28,16 +28,12 @@ public List<MedicineTenderOffer> GetAll()
...
-        {
-            return MedicineTenderOfferRepository.Create(MedicineTenderOfferAdapter.MedicineTenderOfferDtoToMedicineTenderOffer(medicineTenderOfferDto));
-        }
-
         public void CreateAllMedicineTenderOffers(List<MedicineTenderOffer> medicineTenderOffers)
         {
             foreach (MedicineTenderOffer medicineTenderOffer in medicineTenderOffers)
             {
-                Create(MedicineTenderOfferAdapter.MedicineTenderOfferToMedicineTenderOfferDto(new MedicineTenderOffer(medicineTenderOffer.MedicineName, medicineTenderOffer.Quantity, medicineTenderOffer.AvailableQuantity, medicineTenderOffer.Price, PharmacyTenderOfferRepository.getNextTenderPharmacyOfferId())));
+                medicineTenderOffer.PharmacyTenderOfferId = PharmacyTenderOfferRepository.getNextTenderPharmacyOfferId();
```
Comment (trki293) [inline IntegrationWithPharmacies/TenderApi/Service/MedicineTenderOfferService.cs:35]: Correct the method name 'getNextTenderPharmacyOfferId' to start with a capital letter.

File: IntegrationWithPharmacies/UrgentMedicineOrderApi/AbstractFactory/PharmacyGrpcSftp.cs [hunk 12]
```diff
@@ -0,0 +1,52 @@
...
+            if (pharmaciesWithMedicine == null) return null;
+            HttpRequests.UpdateMedicineQuantity(medicine);
+            return ForwardUrgentOrderGrpc(medicine, pharmaciesWithMedicine);
+        }
+
+        public String ForwardUrgentOrderGrpc(string medicine, List<MedicineName> pharmaciesWithMedicine)
+        {
+            UrgentMedicineOrder urgentMedicineOrder = UrgentOrderService.CreateUrgentOrder(medicine, pharmaciesWithMedicine);
+            if (UrgentOrderService.SendOrderGrpc(urgentMedicineOrder)) return CretaeUrgenMedicinetOrder(pharmaciesWithMedicine, urgentMedicineOrder);
+            return null;
```
Comment (trki293) [inline IntegrationWithPharmacies/UrgentMedicineOrderApi/AbstractFactory/PharmacyGrpcSftp.cs:42]: Write this as a ternary operator.
Change name of method 'CretaeUrgenMedicinetOrder' to 'CreateUrgentMedicineOrder'.

File: IntegrationWithPharmacies/UrgentMedicineOrderApi/AbstractFactory/PharmacyHttp.cs [hunk 13]
```diff
@@ -0,0 +1,56 @@
...
+            if (pharmaciesWithMedicine == null) return null;
+            HttpRequests.UpdateMedicineQuantity(medicine);
+            return ForwardUrgentUrderHttp(medicine, pharmaciesWithMedicine);
+        }
+
+        private String ForwardUrgentUrderHttp(String medicine, List<MedicineName> pharmaciesWithMedicine)
+        {
+            UrgentMedicineOrder urgentMedicineOrder = UrgentOrderService.CreateUrgentOrder(medicine, pharmaciesWithMedicine);
+            if (UrgentOrderService.SendOrderHttp(urgentMedicineOrder)) return CretaeUrgenMedicinetOrder(pharmaciesWithMedicine, urgentMedicineOrder);
+            return null;
```
Comment (trki293) [inline IntegrationWithPharmacies/UrgentMedicineOrderApi/AbstractFactory/PharmacyHttp.cs:42]: Write this as a ternary operator and apply a method name change.

File: IntegrationWithPharmacies/UrgentMedicineOrderApi/Service/SmptServerService.cs [hunk 14]
```diff
@@ -0,0 +1,43 @@
...
+using System.Net.Mail;
+using System.Text;
+using UrgentMedicineOrderApi.Model;
+
+namespace UrgentMedicineOrderApi.Service
+{
+    public class SmptServerService
+    {
+        public SmptServerService() { }
+        public void SendEMailNotificationForUrgentOrdee(UrgentMedicineOrder urgentMedicineOrder, string pharmacyName)
```
Comment (trki293) [inline IntegrationWithPharmacies/UrgentMedicineOrderApi/Service/SmptServerService.cs:11]: Change name of method to SendEMailNotificationForUrgentOrder and apply a method name change.

File: IntegrationWithPharmacies/UrgentMedicineOrderApi/Service/SmptServerService.cs [hunk 15]
```diff
@@ -0,0 +1,43 @@
...
+        private static MailMessage CreateMailMessage(UrgentMedicineOrder urgentMedicineOrder, String pharmacyName)
+        {
+            return new MailMessage("ourhospital9@gmail.com", "pharmacyisa@gmail.com", "Notification about urgent medicine order", FormBodyOfMailMessage(urgentMedicineOrder, pharmacyName));
+        }
+
+        private static string FormBodyOfMailMessage(UrgentMedicineOrder urgentMedicineOrder, string pharmacyName)
+        {
+            StringBuilder stringBuilder = new StringBuilder();
+            stringBuilder.Append("Medicine is ordered from: " + pharmacyName + "\n" + "Requested medicine: " + urgentMedicineOrder.Name + ", quantity:  " + urgentMedicineOrder.Quantity + " \n");
+            return stringBuilder.ToString();
```
Comment (trki293) [inline IntegrationWithPharmacies/UrgentMedicineOrderApi/Service/SmptServerService.cs:39]: Write this in form:
"return "Medicine is ordered from: " + pharmacyName + "\n" + "Requested medicine: " + urgentMedicineOrder.Name + ", quantity:  " + urgentMedicineOrder.Quantity + " \n";"

File: IntegrationWithPharmacies/UrgentMedicineOrderApi/Service/UrgentOrderService.cs [hunk 16]
```diff
@@ -5,32 +5,74 @@
...
+        public String FormUrgentOrder(string medicine)
+        {
+            List<String> pharmacies = new List<String>();
+
+            foreach(RegistrationInPharmacy registrationInPharmacy in HttpRequests.GetRegistrationsInPharmaciesAll())
+            {
+                DefineTyepOfApiKey(medicine, pharmacies, registrationInPharmacy);
+            }
+            if (pharmacies.Count!=0) { return pharmacies[0]; }
+            return null;
```
Comment (trki293) [inline IntegrationWithPharmacies/UrgentMedicineOrderApi/Service/UrgentOrderService.cs:50]: Write this as a ternary operator.

**Review by trki293 — CHANGES_REQUESTED** [review]
Modify the indicated places in the code.
Note: check the method names, they should start with a capital letter.

File: IntegrationWithPharmacies/PharmacyRegistrationApi/Repository/RegistrationInPharmacyRepository.cs [hunk 17]
```diff
@@ -27,12 +27,12 @@ public List<RegistrationInPharmacy> GetAll()
        
         public RegistrationInPharmacy getPharmacyApiKey(String apiKey)
```
Comment (trki293) [inline IntegrationWithPharmacies/PharmacyRegistrationApi/Repository/RegistrationInPharmacyRepository.cs:28]: Change name of method to start with upper letter.

File: IntegrationWithPharmacies/MedicineInformationApi/Service/MedicineInformationService.cs [hunk 18]
```diff
@@ -43,13 +43,12 @@ public string GetMedicineDescriptionFromStub(string medicine)
...
         }
 
         public List<MedicineName>  GetAllMedicinesFromDatabase()
         {
-            List<MedicineInformation> medicines = MedicineWithQuantityRepository.GetAll();
-            return getMedicineNames(medicines);
+            return getMedicineNames(MedicineWithQuantityRepository.GetAll());
         }
 
         public List<MedicineName> getMedicineNames(List<MedicineInformation> medicines)
```
Comment (trki293) [inline IntegrationWithPharmacies/MedicineInformationApi/Service/MedicineInformationService.cs:54]: Change name of method to start with upper letter.

File: IntegrationWithPharmacies/UrgentMedicineOrderApi/AbstractFactory/PharmacyGrpcSftp.cs [hunk 19]
```diff
@@ -0,0 +1,51 @@
...
+            HttpRequests.UpdateMedicineQuantity(medicine);
+            return ForwardUrgentOrderGrpc(medicine, pharmaciesWithMedicine);
+        }
+
+        public String ForwardUrgentOrderGrpc(string medicine, List<MedicineName> pharmaciesWithMedicine)
+        {
+            UrgentMedicineOrder urgentMedicineOrder = UrgentOrderService.CreateUrgentOrder(medicine, pharmaciesWithMedicine);
+            return UrgentOrderService.SendOrderGrpc(urgentMedicineOrder) ? CretaeUrgenMedicinetOrder(pharmaciesWithMedicine, urgentMedicineOrder) : null;
+        }
+        private String CretaeUrgenMedicinetOrder(List<MedicineName> pharmaciesWithMedicine, UrgentMedicineOrder urgentMedicineOrder)
```
Comment (trki293) [inline IntegrationWithPharmacies/UrgentMedicineOrderApi/AbstractFactory/PharmacyGrpcSftp.cs:43]: Change method name 'CretaeUrgenMedicinetOrder' to 'CreateUrgentMedicineOrder'.

File: IntegrationWithPharmacies/UrgentMedicineOrderApi/AbstractFactory/PharmacyHttp.cs [hunk 20]
```diff
@@ -0,0 +1,55 @@
...
+            return ForwardUrgentUrderHttp(medicine, pharmaciesWithMedicine);
+        }
+
+        private String ForwardUrgentUrderHttp(String medicine, List<MedicineName> pharmaciesWithMedicine)
+        {
+            UrgentMedicineOrder urgentMedicineOrder = UrgentOrderService.CreateUrgentOrder(medicine, pharmaciesWithMedicine);
+            return UrgentOrderService.SendOrderHttp(urgentMedicineOrder) ? CretaeUrgenMedicinetOrder(pharmaciesWithMedicine, urgentMedicineOrder) : null;
+        }
+
+        private String CretaeUrgenMedicinetOrder(List<MedicineName> pharmaciesWithMedicine, UrgentMedicineOrder urgentMedicineOrder)
```
Comment (trki293) [inline IntegrationWithPharmacies/UrgentMedicineOrderApi/AbstractFactory/PharmacyHttp.cs:44]: Change method name 'CretaeUrgenMedicinetOrder' to 'CreateUrgentMedicineOrder'.

**Review by trki293 — CHANGES_REQUESTED** [review]
Correct the indicated places in the code.

**Review by trki293 — APPROVED** [review]
PR is approved. Good Clean code practices are followed.

**Review by velicko-tam — APPROVED** [review]
PR is approved. Clean code principles were implemented.

**Pull request merged**